pkg/config/migration/multi_account.go

```diff
@@ -0,0 +1,124 @@
+package migration
+
+import (
+	"fmt"
+
+	"github.com/cli/go-gh/v2/pkg/config"
+)
+
+type CowardlyRefusalError struct {
+	Reason string
+}
+
+func (e CowardlyRefusalError) Error() string {
+	// Consider whether we should add a call to action here like "open an issue with the contents of your redacted hosts.yml"
+	return fmt.Sprintf("cowardly refusing to continue with multi account migration: %s", e.Reason)
+}
+
+var hostsKey = []string{"hosts"}
+
+// This migration exists to take a hosts section of the following structure:
+//
+//	github.com:
+//	  user: williammartin
+//	  git_protocol: https
+//	  editor: vim
+//	github.localhost:
+//	  user: monalisa
+//	  git_protocol: https
+//
+// We want this to migrate to something like:
+//
+// github.com:
+//   active_user: williammartin
+//	 users:
+//	   williammartin:
+//	     git_protocol: https
+//	     editor: vim
+//
+// github.localhost:
+//	 active_user: monalisa
+//     users:
+//	     monalisa:
+//		   git_protocol: https
+//
+// The reason for this is that we can then add new users under a host.
+
+type MultiAccount struct{}
+
+func (m MultiAccount) Do(c *config.Config) (bool, error) {
+	hostnames, err := c.Keys(hostsKey)
+	// [github.com, github.localhost]
+	if err != nil {
+		return false, CowardlyRefusalError{"couldn't get hosts configuration"}
+	}
+
+	// If there are no hosts then it doesn't matter whether we migrate or not,
+	// so lets avoid any confusion and say there's no migration required.
+	if len(hostnames) == 0 {
+		return false, nil
+	}
+
+	// If there is an active user set for the first host then we must have already
+	// migrated, so no migration is required. Note that this is a little janky but
+	// without introducing a version into the hosts (which would require special handling elsewhere),
+	// or a version into the config (which would require config backup to move in sync with hosts backup)
+	// it seems like the best option for now.
+	if _, err := c.Get(append(hostsKey, hostnames[0], "active_user")); err == nil {
+		return false, nil
+	}
+
+	// Otherwise let's get to the business of migrating!
+	for _, hostname := range hostnames {
+		configEntryKeys, err := c.Keys(append(hostsKey, hostname))
+		// e.g. [user, git_protocol, editor]
+		if err != nil {
+			return false, CowardlyRefusalError{fmt.Sprintf("couldn't get host configuration despite %q existing", hostname)}
```

Comment:
None of these errors are tested. I also made the decision to return `false` even though a migration was required because it's idiomatic that under error, the other return values are ignored anyway. Didn't want to provide meaning to that value in the case of error. 🤷